### PR TITLE
feat: simple progress report for scheduled transactions

### LINKED ISSUE #245: Simple progress report for scheduled transactions
See https://github.com/joinmarket-webui/joinmarket-webui/pull/242.

Add a simple progress report based on the transaction schedule to let the user know how long a schedule is roughly going to last and what its progress is.

### PR DESCRIPTION
**Pull request by dnlggr**
Resolves #245.

~Still need to fix a few smaller things like i18n and coordinating the requests and placeholders on initial page load.~

As always, the copy is just off the top of my head. Feel free to change it!

But in principle it should be ready for review I hope. The CSS looks more complicated than it actually is. Sorry for that. I basically adapted an existing progress bar I found for our use case and it just grew into what it is now. 😅 

Let me know what you think!

## 📸 

https://user-images.githubusercontent.com/10026790/168109364-f609cd7e-3b3d-4bcb-8cbd-d37b4ca1083b.mov

### CODE REVIEW

**Review by editwentyone — COMMENTED** [review]
already a very good start! I like it so far. is there a need to add the possibility to read/view "logs" live?

**Review by MaxHillebrand — APPROVED** [review]
This is great!

Why are the points on the time line not evenly distributed?

**Comment by dnlggr** [comment]
> Why are the points on the time line not evenly distributed?

The points represent transactions. The schedule includes randomly chosen wait times between each transaction (to prevent timing attacks). The distance between two points roughly corresponds to the wait time between the two respective transactions. Each schedule run will have a different distribution of points on the time line.

**Comment by dnlggr** [comment]
Should be ready now @theborakompanioni. 🙏 Sorry for forgetting to actually reload the schedule first. 😅

File: src/components/Jam.jsx [hunk 1]
```diff
@@ -106,22 +147,21 @@ export default function Jam() {
...
-    //     liquiditywait: 10,
-    //     waittime: 1.0,
-    //     mixdepthsrc: 0,
-    //   }
-    // }
+    if (process.env.NODE_ENV === 'development' && useInsecureTestingSettings && !deactivateTestingToggle) {
+      body.tumbler_options = {
+        addrcount: 3,
+        minmakercount: 1,
+        makercountrange: [1, 1],
```
Comment (theborakompanioni) [inline src/components/Jam.jsx:154]:
```suggestion
        makercountrange: [1, 0],
```

https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/50f5228498c670258740cd3cab485bcb39e177a3/jmclient/jmclient/cli_options.py#L317-L318

Second param is `spread`. We want exactly one maker in regtest.

File: src/components/Jam.jsx [hunk 2]
```diff
@@ -72,6 +77,42 @@ export default function Jam() {
     setCollaborativeOperationRunning(coinjoinInProgress || makerRunning)
   }, [serviceInfo])
 
+  const reloadSchedule = useCallback(
+    ({ signal }) => {
+      return Api.getTumblerSchedule({ walletName: wallet.name, token: wallet.token })
```
Comment (theborakompanioni) [inline src/components/Jam.jsx:82]:
```suggestion
      return Api.getTumblerSchedule({ walletName: wallet.name, token: wallet.token, signal })
```

`signal` is unused.

File: src/components/Jam.jsx [hunk 3]
```diff
@@ -72,6 +77,42 @@ export default function Jam() {
...
+        .then((res) => (res.ok ? res.json() : Api.Helper.throwError(res)))
+        .then((data) => {
+          setSchedule(data.schedule)
```
Comment (theborakompanioni) [inline src/components/Jam.jsx:85]:
```suggestion
          !signal.aborted && setSchedule(data.schedule)
```

File: src/components/Jam.jsx [hunk 4]
```diff
@@ -72,6 +77,42 @@ export default function Jam() {
...
+        })
+        .catch((err) => {
+          // Not finding a schedule is not an error.
+          // It means a single colalborative transaction is running.
+          // Those have no schedule.
```
Comment (theborakompanioni) [inline src/components/Jam.jsx:90]: Not finding a schedule with 404 is not an error. Everything else is an error, isn't it?
Status code can be obtained with `err.response.status` if it is an `JmApiError`.

**Review by theborakompanioni — APPROVED** [review]
Looks quite nice overall.

I really laughed out loud when I saw "finished in 53 seconds" and the corresponding translation string `progress_tldr_seconds` – shows how you care about little details, even if very few people will ever see it.

What helped reviewing was to see the raw schedule data, so it would be nice to log it to console or even display it below the form in developer mode.

I'll spare you my nits, most of them can be addressed when working on the other related issues.

Other than that and what was mentioned as comment: Again, really nice. Approved :rocket:

File: src/components/Jam.jsx (hunk 4 above)
Comment (dnlggr) [reply src/components/Jam.jsx]: Good point! Updated. ✅

**Comment by dnlggr** [comment]
> even if very few people will ever see it.

No one will ever see it, all prod schedules will be way longer. But we will see it every single time while testing. Worth it. 😄 

> I'll spare you my nits, most of them can be addressed when working on the other related issues.

Don't worry, let's hear the nits! We can always delay addressing them anyway.

**Pull request merged**